drivers/dac/ad5754r/ad5754r.c

```diff
@@ -0,0 +1,767 @@
+ */
+int ad5754r_set_int_ref_pwrup(struct ad5754r_dev *dev,
+			      enum ad5754r_pwr_int_ref_pwrup int_ref_pwrup)
+{
+	int ret;
+
+	if (int_ref_pwrup > AD5754R_PWR_INT_REF_POWERUP)
+		return -EINVAL;
+
+	// TODO Confirm with apps on whether the DAC channel address
+	// is don't care, or 0
+
+	ret = ad5754r_update_bits(dev,
+				  AD5754R_PREP_INSTR_ADDR(AD5754R_REG_PWR_CTRL, 0),
+				  AD5754R_PWR_UP_INT_REF_MASK,
+				  int_ref_pwrup);
+	if (ret)
+		return ret;
+
+	dev->int_ref_pwrup = int_ref_pwrup;
+
+	return 0;
+}
+
+/**
+ * @brief Get Over-current status bit for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param oc_ch_alert - Store channel overcurrent alert status.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_get_oc_ch_alert(struct ad5754r_dev *dev, uint8_t chn,
+			    enum ad5754r_pwr_oc_ch_alert *oc_ch_alert)
+{
+	int ret;
+	uint16_t reg_val;
+
+	if (!oc_ch_alert || (chn >= AD5754R_NUM_CHANNELS))
+		return -EINVAL;
+
+	// TODO Confirm whether the DAC channel address
+	// is don't care, or 0
+
+	ret = ad5754r_read(dev, AD5754R_PREP_INSTR_ADDR(AD5754R_REG_PWR_CTRL, 0),
+			   &reg_val);
+	if (ret)
+		return ret;
+
+	*oc_ch_alert = no_os_field_get(AD5754R_PWR_OC_ALERT_CH_MASK(chn), reg_val);
+
+	return 0;
+}
+
+/**
+ * @brief Get Thermal Shutdown status bit.
+ * @param dev - The device structure.
+ * @param tsd_alert - Store TSD alert status.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_get_tsd_alert(struct ad5754r_dev *dev,
+			  enum ad5754r_pwr_tsd_alert *tsd_alert)
+{
+	int ret;
+	uint16_t reg_val;
+
+	if (!tsd_alert)
+		return -EINVAL;
+
+	// TODO Confirm whether the DAC channel address
+	// is don't care, or 0
+
+	ret = ad5754r_read(dev,
+			   AD5754R_PREP_INSTR_ADDR(AD5754R_REG_PWR_CTRL, 0),
+			   &reg_val);
+	if (ret)
+		return ret;
+
+	*tsd_alert = no_os_field_get(AD5754R_PWR_TSD_ALERT_MASK, reg_val);
+
+	return 0;
+}
+
+/**
+ * @brief Write to and read back a register to verify SPI.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_spi_verify(struct ad5754r_dev *dev)
+{
+	int ret;
+	uint16_t reg_data;
+	uint8_t instr_addr;
+
+	if (!dev)
+		return -EINVAL;
+
+	/* Use Output Range Select register for DAC Channel A, to verify SPI */
+	instr_addr = AD5754R_PREP_INSTR_ADDR(AD5754R_REG_OUTPUT_RANGE_SEL,
+						AD5754R_DAC_CH_A_ADDR);
+
+	/* Change Channel A output range to 0-10V */
+	ret = ad5754r_set_ch_range(dev, AD5754R_DAC_CH_A, AD5754R_SPAN_0V_TO_10V);
+	if (ret)
+		return ret;
+
+	/* Read channel A output range */
+	ret = ad5754r_read(dev, instr_addr, &reg_data);
+	if (ret)
+		return ret;
+
+	/* Confirm if readback value matches what was written */
+	if (no_os_field_get(AD5754R_OUTPUT_RANGE_SEL_MASK,
+			    reg_data) != AD5754R_SPAN_0V_TO_10V)
+		return -EINVAL;
```

Comment:
You can use -EBADMSG here instead of -EINVAL. EINVAL is more for wrong set of input/s.